logstash-core/src/main/java/org/logstash/plugins/codecs/Line.java

```diff
@@ -52,12 +52,8 @@
 
     private final CharBuffer charBuffer = ByteBuffer.allocateDirect(64 * 1024).asCharBuffer();
     private final CharsetDecoder decoder;
```

Comment:
I'm still worried by `CharsetDecoder` being an attribute of `Codec`, since it is internally stateful.

Reply:
On the input side, there is only a single codec so they can (and often must) be stateful across calls to `decode()`.

Reply:
Ah. So we provide a single instance of `Codec` to either `Input` or `Output`, and the cautionary tale in docs is largely a reminder that a single instance of an `Output` (and therefore its single instance of `Codec`) will be shared across threads when `pipeline.workers` GT 1, so internal statefulness matters in implementation 

Do we provide a way for inputs to get a new instance of the specified codec so they can avoid sharing something that is inherently stateful? I see this being necessary with inputs like UDP/TCP where an input is tasked with reading multiple inbound streams of data simultaneously.

Reply:
Yep, that's what [cloneCodec()](https://github.com/elastic/logstash/blob/nsjp_ga/logstash-core/src/main/java/co/elastic/logstash/api/Codec.java#L53-L59) is intended to enable.